### DISCUSSION TITLE
How do I add a class to all links?

### DISCUSSION THREAD

**Discussion by ctessmer** [thephpleague/commonmark#1031] [Q&A] [answered]
# What I want

I'd like to add a class to all my links. For instance, I'd like to convert this:

#### Markdown
```markdown
What has [myspace](https://myspace.com) become?!
```

To this:

#### HTML
```html
What has <a href="https://myspace.com" class="link">myspace</a> become?!
```

# Why I want it

I use tailwindcss and have some nice components setup—one being a `link` component. I'd like to attach that to every link in the rendered html so I don't have to duplicate the component in css. 

#### Tailwindcss component
```css
.link {
     @apply underline decoration-2 text-blue-600 hover:text-blue-800 dark:text-blue-400 font-semibold;
}
```

# What I tried

It looks like there's something perhaps built-in to do this at the end of the [Abstract Syntax Tree section](https://commonmark.thephpleague.com/2.4/customization/abstract-syntax-tree/#html-attributes):

#### HTML Attributes
```php
use League\CommonMark\Extension\CommonMark\Node\Inline\Link;

$link = new Link('https://twitter.com/colinodell', '@colinodell');
$link->data->append('attributes/class', 'social-link');
$link->data->append('attributes/class', 'twitter');
$link->data->set('attributes/target', '_blank');
$link->data->set('attributes/rel', 'noopener');
```

But I'm unsure how to…

1. Integrate the above `HTML Attributes` example into my code (just to get a proof of concept).
2. Adapt it so it targets *all* links—not just text that says `@colinodell`.
3. Register into my converter (not sure if register is the right word).

# My existing, working conversion code

```php
# Use markdown converter library
use League\CommonMark\Environment\Environment;
use League\CommonMark\Extension\CommonMark\CommonMarkCoreExtension;
use League\CommonMark\Extension\SmartPunct\SmartPunctExtension;
use League\CommonMark\Extension\Strikethrough\StrikethroughExtension;
use League\CommonMark\Extension\FrontMatter\FrontMatterExtension;
use League\CommonMark\Extension\FrontMatter\Output\RenderedContentWithFrontMatter;
use League\CommonMark\MarkdownConverter;

# Define markdown converter environment
$environment = new Environment();
$environment->addExtension(new CommonMarkCoreExtension());
$environment->addExtension(new SmartPunctExtension());
$environment->addExtension(new StrikethroughExtension());
$environment->addExtension(new FrontMatterExtension());

# Define converter
$converter = new MarkdownConverter($environment);
```

**Can you tell me how to solve this?**

PS: Thanks so much for the incredible library. It still blows me away that I can use it (and others' libraries) for free.

**Comment by colinodell** [comment, marked as answer]
The [Default Attributes extension](https://commonmark.thephpleague.com/2.4/extensions/default-attributes/#usage) does exactly this - no custom code required!

If you really did want to implement this with custom code, [here's an example you could easily adapt](https://commonmark.thephpleague.com/2.4/customization/event-dispatcher/#example) - just change this part:

```diff
     public function onDocumentParsed(DocumentParsedEvent $event): void
     {
         $document = $event->getDocument();
         $walker = $document->walker();
         while ($event = $walker->next()) {
             $node = $event->getNode();
 
             // Only stop at Link nodes when we first encounter them
             if (!($node instanceof Link) || !$event->isEntering()) {
                 continue;
             }

-            $url = $node->getUrl();
-            if ($this->isUrlExternal($url)) {
-                $node->data->append('attributes/class', 'external-link');
-            }
+            $link->data->append('attributes/class', 'link');
         }
     }
```

But since that extension should do everything you need I'd recommend just using that :)

> PS: Thanks so much for the incredible library. It still blows me away that I can use it (and others' libraries) for free.

Thank you so much for those kind words!

**Comment by ctessmer** [comment]
> The [Default Attributes extension](https://commonmark.thephpleague.com/2.4/extensions/default-attributes/#usage) does exactly this - no custom code required!

Perfect! Thank you so much for letting me know. Exactly what I was looking for. (And I thank you for going the extra mile and showing me how to do it in a totally custom way!)

And btw I tried it, and it *almost* works like a charm—but I may have discovered a bug (one that I can easily work around). But just so you know:

# Bug overview

The library throws a fatal error and exits if the value of the class is `link`.

# How to duplicate

1. Set config:

```php
$config = [
	'default_attributes' => [
		Link::class => [
			'class' => 'link',
			'target' => '_self',
		],
	],
];
```

2. Supply library with a link in markdown:

#### Markdown
```markdown
What has [myspace](https://myspace.com) become?!
```

3. Execute code.

# Expected result

#### HTML
```html
<p>What has <a href="https://myspace.com" class="link">myspace</a> become?!</p>
```

# Actual result

`Fatal error: Uncaught ArgumentCountError: link() expects exactly 2 arguments, 1 given…`
